### DISCUSSION TITLE
Suggestion on modeling a modal window

### DISCUSSION THREAD

**Discussion by coodoo** [statelyai/xstate#2051] [Q&A]
Say we have got a typical `master/detail` page setup, with a master page showing a list of items, when one of them was clicked, goes into a detail page showing the content of that item, also on that detail page is a `remove item` button, when clicked, showing a `modal` window with confirmation texts and two buttons: `Cancel` and `Delete`.

Currently I'm modeling this with parallel state as seen below, the issue here is, if `Delete` was clicked, we need to go back to the `master` page, if `Cancel` clicked, we need to stay in current `detail` page, but obviously inter `parallel` state is not allowed, hence I'm wondering how would go about it?

Or precisely, is there a recommended way of modeling `modal` or [`notification` style](https://github.com/fkhadra/react-toastify) kind of window?

```js
{
  type: 'parallel',
  states: {
    main: {
      initial: 'loading',
      states: {
        loading: {...},
        master: {...},
        detail: {...},
      }

    },
    modal: {...},
  }
}
```

**Comment by davidkpiano** [comment]
Great question! This is one of the fun (creative) part of state machines/statecharts - thinking about how you can better model use-cases like this.

Having `modal` be in an orthogonal state might not make sense because, as you said, it really only exists as part of the `detail` state. So, we can make it a nested state instead:

```js
const machine = Machine({
  // ...
  states: {
    loading: { ... },
    master: { ... },
    detail: {
      initial: 'idle', // nothing happening
      states: {
        idle: {
          on: {
            MAYBE_DELETE: 'deleteConfirm'
          },
        },
        deleteConfirm: {
          on: {
            CANCEL: 'idle'
          }
        }
      },
      on: {
        DELETE: {
          target: 'master',
          actions: 'deleteItem'
        }
      }
    }
  }
});
```

This has the feature of transitions for two events: `MAYBE_DELETE` for prompting a confirmation, or `DELETE` for deleting without a confirmation.

**Comment by coodoo** [comment]
Ah I didn't make the question clear enough, there's actually two issues here:

### 1. The modal window could be triggered by any of the states

Say in `loading` state if network failed it will display the error message in the modal window, and for `master` state, let's say each row in the listing has a `delete` button, when clicked, it will also trigger the modal window.

Hence the issue comes down to: how to share the modal window(state) between all states? and better yet, let them communicate?

### 2. The modal window will be overlaid on top of the detail screen

Which means there will be two views displayed at the same time (detail below, modal on top), and each of them belongs to different state, that's why I tried to model it with parallel state.

<img width="461" alt="Screen Shot 2019-04-12 at 8 52 25 AM" src="https://user-images.githubusercontent.com/325936/56005067-33633180-5d01-11e9-9e5b-1ee222c98162.png">

**Comment by coodoo** [comment]
cc @johnyanarella :)

**Comment by JosephusPaye** [comment]
@coodoo Did you find a good way to model this?
